src/widgets/dialogs/RestoreBackupsDialog.cpp

```diff
@@ -0,0 +1,159 @@
+    , fileData(std::move(fileData))
+    , backupCombo(new QComboBox)
+    , showButton(u"Show"_s)
+    , corruptedBackupsWarning(
+          u"Some backups are damaged or otherwise unreadable."_s)
+{
+    this->setAttribute(Qt::WA_DeleteOnClose);
+    this->setWindowTitle(u"Restore backup of " %
+                         this->fileData.fileKind.toLower() % '?');
+
+    auto *layout = new QVBoxLayout(this);
+
+    auto *description =
+        new QLabel(this->fileData.fileKind % u" failed to load: " % prevError %
+                   u"<p>" % this->fileData.fileDescription %
+                   u"<p>There are backups of this file.<br>Do you want to "
+                   u"restore the selected backup?");
+    layout->addWidget(description);
+
+    auto *hbox = new QHBoxLayout;
+    hbox->addWidget(&this->backupCombo, 1);
+    hbox->addWidget(&this->showButton);
+    layout->addLayout(hbox);
+
+    this->corruptedBackupsWarning.hide();
+    layout->addWidget(&this->corruptedBackupsWarning);
+
+    layout->addSpacing(10);
+
+    auto *buttons =
+        new QDialogButtonBox(QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
+    layout->addWidget(buttons);
+
+    QObject::connect(buttons, &QDialogButtonBox::accepted, this, [this] {
+        auto selected = this->backupCombo.currentData();
+        auto *data = get_if<backup::BackupFile>(&selected);
+        if (!data)
+        {
+            return;
+        }
+
+        bool retry = true;
+        while (retry)
+        {
+            retry = false;
+            std::error_code ec;
+            qCDebug(chatterinoSettings)
+                << "Copying" << stdPathToQString(data->path) << "to"
+                << stdPathToQString(data->dstPath);
+            if (!std::filesystem::copy_file(
+                    data->path, data->dstPath,
+                    std::filesystem::copy_options::overwrite_existing, ec))
+            {
+                retry = QMessageBox::critical(
+                            this, "Failed to restore file",
+                            u"Failed to copy '%1' to '%2': %3"_s.arg(
+                                stdPathToQString(data->path),
+                                stdPathToQString(data->dstPath),
+                                QString::fromStdString(ec.message())),
+                            QMessageBox::Retry | QMessageBox::Ok) ==
+                        QMessageBox::Retry;
+            }
+        }
+
+        this->accept();
+    });
+    QObject::connect(buttons, &QDialogButtonBox::rejected, this,
+                     &QDialog::reject);
+
+    QObject::connect(&this->showButton, &QPushButton::clicked, this, [this] {
+        auto selected = this->backupCombo.currentData();
+        auto *data = get_if<backup::BackupFile>(&selected);
+        if (!data)
+        {
+            return;
+        }
+        auto url = QUrl::fromLocalFile(stdPathToQString(data->path));
+        QDesktopServices::openUrl(url);
+    });
+
+    this->refreshBackups();
```

Comment:
It would be good to set a fixed size for this dialog - that way it will properly show up as a popup on sway

this worked for me and replicates what QMessageBox uses:
```
    this->layout()->activate();
    this->setFixedSize(this->layout()->totalMinimumSize());
```